File: server/routes/smartMatch.js
const express = require('express');
const User = require('../models/User');
const Project = require('../models/Project');
const auth = require('../middleware/auth');
const { matchUserToProject, optimizeProfile } = require('../utils/gemini');

const router = express.Router();

// Get smart matches based on skills
router.get('/skills', auth, async (req, res) => {
  try {
    const currentUser = await User.findById(req.user._id);
    const userSkills = currentUser.profile?.skills || [];
    
    if (userSkills.length === 0) {
      return res.json([]);
    }

    // Find users with similar skills
    const matches = await User.find({
      _id: { $ne: req.user._id },
      'profile.skills': { $in: userSkills }
    })
    .select('-password')
    .limit(10);

    res.json(matches);
  } catch (error) {
    console.error('Smart match error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// Get smart matches based on goals
router.get('/goals', auth, async (req, res) => {
  try {
    const currentUser = await User.findById(req.user._id);
    const userType = currentUser.userType;
    
    // Find users with similar career goals (same user type)
    const matches = await User.find({
      _id: { $ne: req.user._id },
      userType: userType
    })
    .select('-password')
    .limit(10);

    res.json(matches);
  } catch (error) {
    console.error('Goal match error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// Get smart matches for project collaboration
router.get('/projects', auth, async (req, res) => {
  try {
    const currentUser = await User.findById(req.user._id);
    const userSkills = currentUser.profile?.skills || [];
    
    // Find users who might be good for collaboration
    const matches = await User.find({
      _id: { $ne: req.user._id },
      $or: [
        { 'profile.skills': { $in: userSkills } },
        { userType: { $ne: currentUser.userType } } // Different user types for diverse collaboration
      ]
    })
    .select('-password')
    .limit(10);

    res.json(matches);
  } catch (error) {
    console.error('Project match error:', error);
    res.status(500).json({ message: 'Server error' });
  }
});

// AI Analyze match between user and a specific project
router.post('/ai-analyze', auth, async (req, res) => {
  try {
    const { projectId } = req.body;
    const project = await Project.findById(projectId);
    const user = await User.findById(req.user._id);

    if (!project || !user) {
      return res.status(404).json({ message: 'Project or User not found' });
    }

    const analysis = await matchUserToProject(user, project);
    res.json(analysis);
  } catch (error) {
    console.error('AI Analysis error:', error);
    res.status(500).json({ message: 'AI Analysis failed' });
  }
});

// AI Optimize user profile
router.post('/optimize-profile', auth, async (req, res) => {
  try {
    const user = await User.findById(req.user._id);
    const optimization = await optimizeProfile(user.profile);
    
    if (!optimization) {
      return res.status(500).json({ message: 'Optimization failed' });
    }

    res.json(optimization);
  } catch (error) {
    console.error('AI Profile Optimization error:', error);
    res.status(500).json({ message: 'Profile optimization failed' });
  }
});

module.exports = router;

File: server-python/routes/smartMatch.py
from flask import Blueprint, request, jsonify
from models import User, Project
from extensions import db
from routes.middleware import token_required
import json
import os

smart_match_bp = Blueprint('smart_match', __name__)

# Note: Integration with google-generativeai for AI analysis would go here.
# For now we'll implement the basic matching logic.

@smart_match_bp.route('/skills', methods=['GET'])
@token_required
def get_skill_matches(current_user):
    try:
        profile = current_user.get_profile()
        user_skills = profile.get('skills', [])
        
        if not user_skills:
            return jsonify([]), 200
            
        # Simplistic matching: Find users who have at least one common skill
        # In a real app, this would be more complex (SQL join or full-text)
        # For SQLite, we search for the skill strings in profile_data
        matches = []
        for s in user_skills:
            users = User.query.filter(User.id != current_user.id, User.profile_data.ilike(f"%{s}%")).limit(10).all()
            for u in users:
                if u.id not in [m['_id'] for m in matches]:
                    matches.append(u.to_dict())
            if len(matches) >= 10:
                break
                
        return jsonify(matches[:10]), 200
    except Exception as e:
        return jsonify({"message": "Server error", "error": str(e)}), 500

@smart_match_bp.route('/goals', methods=['GET'])
@token_required
def get_goal_matches(current_user):
    try:
        matches = User.query.filter(User.id != current_user.id, User.user_type == current_user.user_type).limit(10).all()
        return jsonify([u.to_dict() for u in matches]), 200
    except Exception as e:
        return jsonify({"message": "Server error", "error": str(e)}), 500

@smart_match_bp.route('/projects', methods=['GET'])
@token_required
def get_project_collaboration_matches(current_user):
    try:
        # Matches users with different user types for diverse collaboration
        matches = User.query.filter(User.id != current_user.id, User.user_type != current_user.user_type).limit(10).all()
        return jsonify([u.to_dict() for u in matches]), 200
    except Exception as e:
        return jsonify({"message": "Server error", "error": str(e)}), 500

@smart_match_bp.route('/ai-analyze', methods=['POST'])
@token_required
def ai_analyze(current_user):
    # This would call matchUserToProject(user, project) using Gemini
    return jsonify({
        "score": 85,
        "summary": "Great match based on project requirements and your profile skills.",
        "pros": ["Strong React.js experience", "Previous project success"],
        "cons": ["Location difference", "Budget slightly below expectation"]
    }), 200

@smart_match_bp.route('/optimize-profile', methods=['POST'])
@token_required
def optimize_profile(current_user):
    # This would call optimizeProfile(user.profile) using Gemini
    return jsonify({
        "summary": "Your profile is 70% complete. Adding more detailed project descriptions will improve match rates.",
        "suggestions": ["Add a headline", "Detail your role in Project X", "List current certifications"]
    }), 200
